src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -293,6 +433,132 @@ internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projec
+                    // are removing the project reference A->B:
+                    //
+                    //   A -> B -> D
+                    //   A -> D
+                    //
+                    // For this example, we do not hit this optimization because D contains project B in the set of
+                    // reverse transitive references. Without more complicated checks, we cannot rule out the
+                    // possibility that A may have been removed from the reverse transitive references of D by the
+                    // removal of the A->B reference.
+                    continue;
+                }
+
+                if (references.Contains(projectId))
+                {
+                    builder.Remove(project);
+                }
+            }
+
+            // Invalidate the transitive references from the previously-referenced project
+            builder.Remove(referencedProjectId);
+
+            return builder.ToImmutable();
+        }
+
+        internal ProjectDependencyGraph WithProjectRemoved(ProjectId projectId)
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+
+            // Project ID set and direct forward references are trivially updated by removing the key corresponding to
+            // the project getting removed.
+            var projectIds = _projectIds.Remove(projectId);
+            var referencesMap = ComputeNewReferencesMapForRemovedProject(_lazyReverseReferencesMap, _referencesMap, projectId);
+
+            // The direct reverse references map is updated by removing the key for the project getting removed, and
+            // also updating any direct references to the removed project.
+            var reverseReferencesMap = _lazyReverseReferencesMap is object
+                ? ComputeNewReverseReferencesMapForRemovedProject(_referencesMap, _lazyReverseReferencesMap, projectId)
+                : null;
+            var transitiveReferencesMap = ComputeNewTransitiveReferencesMapForRemovedProject(_transitiveReferencesMap, projectId);
+            var reverseTransitiveReferencesMap = ComputeNewReverseTransitiveReferencesMapForRemovedProject(_reverseTransitiveReferencesMap, projectId);
+            return new ProjectDependencyGraph(
+                projectIds,
+                referencesMap,
+                reverseReferencesMap,
+                transitiveReferencesMap,
+                reverseTransitiveReferencesMap,
+                topologicallySortedProjects: default,
+                dependencySets: default);
+        }
+
+        internal ProjectDependencyGraph WithProjectReferenceRemoved(ProjectId projectId, ProjectId referencedProjectId)
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+            Contract.ThrowIfFalse(_referencesMap[projectId].Contains(referencedProjectId));
+
+            var projectIds = _projectIds;
```

Comment:
consider doc'ing these steps.

Reply:
The current code reads "update each property according to the current value and current operation". Is there a specific part that is not clear?

Reply:
i actually find it less clear to have this local as i thought it was going to be used for something.  i think that threw me off.  i think i'd prefer to just see the updates to only what has changed.  thus knowing what was out of scope for hte change.  but it's not major and the code is still reasonably clear :)

Reply:
Updated comments to clarify this